File: rb/test/highlight_entity_test.rb
# Highlight entity test

require "minitest/autorun"
require "json"
require_relative "../SvobodaWeb_sdk"
require_relative "runner"

class HighlightEntityTest < Minitest::Test
  def test_create_instance
    testsdk = SvobodaWebSDK.test(nil, nil)
    ent = testsdk.Highlight(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "highlight" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = SvobodaWebSDK.test(seed, nil)
    seen = base.Highlight(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = SvobodaWebConfig.shared_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = SvobodaWebSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.Highlight(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = highlight_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "highlight." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set SVOBODA_WEB_TEST_HIGHLIGHT_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    highlight_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.highlight")))
    highlight_ref01_data = nil
    if highlight_ref01_data_raw.length > 0
      highlight_ref01_data = Helpers.to_map(highlight_ref01_data_raw[0][1])
    end

    # LIST
    highlight_ref01_ent = client.Highlight(nil)
    highlight_ref01_match = {}

    highlight_ref01_list_result = highlight_ref01_ent.list(highlight_ref01_match, nil)
    assert highlight_ref01_list_result.is_a?(Array)

  end
end

def highlight_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "highlight", "HighlightTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = SvobodaWebSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["highlight01", "highlight02", "highlight03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["SVOBODA_WEB_TEST_HIGHLIGHT_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "SVOBODA_WEB_TEST_HIGHLIGHT_ENTID" => idmap,
    "SVOBODA_WEB_TEST_LIVE" => "FALSE",
    "SVOBODA_WEB_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["SVOBODA_WEB_TEST_HIGHLIGHT_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["SVOBODA_WEB_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = SvobodaWebSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["SVOBODA_WEB_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["SVOBODA_WEB_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: ts/test/entity/highlight/HighlightEntity.test.ts


import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'
import { createLiveTransport } from '../../live-runner'
import { runLiveEntity } from '../../live-entity'


import { SvobodaWebSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveClientOptions,
  liveDelay,
  loadEnvLocal,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


// AFTER the imports on purpose: TypeScript hoists `import` above any
// statement in the emitted CommonJS, so a loader placed above them would
// run only after every imported module had already been evaluated - and
// anything reading process.env at module scope would miss these values.
loadEnvLocal(__dirname + '/../../../.env.local')


describe('HighlightEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when SVOBODA_WEB_TEST_LIVE=TRUE.
  afterEach(liveDelay('SVOBODA_WEB_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = SvobodaWebSDK.test()
    const ent = testsdk.Highlight()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.SVOBODA_WEB_TEST_LIVE
    for (const op of ['list']) {
      if (!live && maybeSkipControl(t, 'entityOp', 'highlight.' + op, live)) return
    }

    
    const setup = basicSetup()
    if (setup.live) {
      return runLiveEntity(setup, {"active":true,"alias":{"field":{}},"fields":[{"active":true,"name":"content","req":false,"short":"Content or description of the highlight","type":"`$STRING`","index$":0},{"active":true,"name":"id","req":false,"short":"Unique identifier for the highlight","type":"`$STRING`","index$":1},{"active":true,"format":"date-time","name":"timestamp","req":false,"short":"Timestamp when the highlight was created or updated","type":"`$STRING`","index$":2},{"active":true,"name":"title","req":false,"short":"Title of the highlight","type":"`$STRING`","index$":3},{"active":true,"format":"uri","name":"url","req":false,"short":"URL to the full article or content","type":"`$STRING`","index$":4}],"id":{"field":"id","name":"id"},"name":"highlight","op":{"list":{"input":"data","name":"list","points":[{"active":true,"args":{},"contract":{"id":"GET /hljson","json":"{\"operationId\":\"getLiveHighlights\",\"parameters\":[],\"protocol\":\"http\",\"responses\":{\"200\":{\"content\":{\"application/json\":{\"example\":{\"highlights\":[{\"content\":\"Latest developments in uncensored news coverage\",\"id\":\"12345\",\"timestamp\":\"2024-01-15T12:30:00Z\",\"title\":\"Breaking News Update\",\"url\":\"https://www.svoboda.org/article/12345\"}]},\"schema\":{\"properties\":{\"highlights\":{\"description\":\"Array of live highlight items\",\"items\":{\"properties\":{\"content\":{\"description\":\"Content or description of the highlight\",\"type\":\"string\"},\"id\":{\"description\":\"Unique identifier for the highlight\",\"type\":\"string\"},\"timestamp\":{\"description\":\"Timestamp when the highlight was created or updated\",\"format\":\"date-time\",\"type\":\"string\"},\"title\":{\"description\":\"Title of the highlight\",\"type\":\"string\"},\"url\":{\"description\":\"URL to the full article or content\",\"format\":\"uri\",\"type\":\"string\"}},\"type\":\"object\"},\"type\":\"array\"}},\"type\":\"object\"}}},\"description\":\"Successful response with live highlights data\"},\"500\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"error\":{\"description\":\"Error message\",\"type\":\"string\"}},\"type\":\"object\"}}},\"description\":\"Internal server error\"}},\"securitySource\":\"unspecified\"}","source":"openapi3","version":1},"kind":"http","method":"GET","orig":"/hljson","segments":[{"lit":"hljson"}],"select":{},"transform":{"req":"`reqdata`","res":"`body.highlights`"},"index$":0}],"key$":"list"}},"relations":{"ancestors":[]},"key$":"highlight","name__orig":"highlight","Name":"Highlight","name_":"highlight","name-":"highlight","NAME":"HIGHLIGHT","index$":0}, {"active":true,"entity":"highlight","key$":"BasicHighlightFlow","kind":"basic","name":"BasicHighlightFlow","param":{},"step":[{"active":true,"data":{},"input":{},"match":{},"op":"list","spec":[],"valid":[{"apply":"ItemExists","def":{"ref":"highlight_ref01"}}],"index$":0}]}, 'Highlight')
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let highlight_ref01_data = Object.values(setup.data.existing.highlight)[0] as any

    // LIST
    const highlight_ref01_ent = client.Highlight()
    const highlight_ref01_match: any = {}

    const highlight_ref01_list = (await highlight_ref01_ent.list(highlight_ref01_match)).map((e: any) => e.data())


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/highlight/HighlightTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = SvobodaWebSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['highlight01','highlight02','highlight03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'SVOBODA_WEB_TEST_HIGHLIGHT_ENTID': idmap,
    'SVOBODA_WEB_TEST_LIVE': 'FALSE',
    'SVOBODA_WEB_TEST_EXPLAIN': 'FALSE',
  })

  idmap = env['SVOBODA_WEB_TEST_HIGHLIGHT_ENTID']

  const live = 'TRUE' === env.SVOBODA_WEB_TEST_LIVE

  const transport = createLiveTransport()
  if (live) {
    const rawIds = process.env['SVOBODA_WEB_TEST_HIGHLIGHT_ENTID']
    idmap = rawIds && rawIds.trim() ? JSON.parse(rawIds) : {}
    if (!idmap || Array.isArray(idmap) || typeof idmap !== 'object') {
      throw new Error('Live ENTID must be a JSON object')
    }
    client = new SvobodaWebSDK(merge([
      // FIRST, so the generated fields below win: sdk-test-control.json's
      // test.client.options adds to the live client, it does not redirect it.
      liveClientOptions(),
      {
      },
      // 'extra || {}', not a bare 'extra': struct.merge returns UNDEFINED when the
      // last entry is undefined, and basicSetup is normally called with no
      // argument at all - so a bare 'extra' silently discarded the apikey
      // and server values above and handed the SDK undefined. Harmless
      // while there was nothing in that object; not harmless now.
      extra || {},
      { system: { fetch: transport.fetch } }
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.SVOBODA_WEB_TEST_EXPLAIN,
    live,
    transport,
    now: Date.now(),
  }

  return setup
}
  
